cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,697 @@
+                                                          ArrowArray* out);
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::dictionary32>(cudf::column& column,
+                                                             rmm::cuda_stream_view stream,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             ArrowArray* out);
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::struct_view>(cudf::column& column,
+                                                            rmm::cuda_stream_view stream,
+                                                            rmm::mr::device_memory_resource* mr,
+                                                            ArrowArray* out)
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(ArrowArrayInitFromType(tmp.get(), NANOARROW_TYPE_STRUCT));
+  NANOARROW_RETURN_NOT_OK(ArrowArrayAllocateChildren(tmp.get(), column.num_children()));
+  tmp->length     = column.size();
+  tmp->null_count = column.null_count();
+
+  auto contents = column.release();
+  if (contents.null_mask) {
+    NANOARROW_RETURN_NOT_OK(
+      set_buffer(std::move(contents.null_mask), kValidityBufferIdx, tmp.get()));
+  }
+
+  ArrowArray* child_ptr = tmp->children[0];
+  for (auto&& child : contents.children) {
+    if (child->type().id() == cudf::type_id::EMPTY) {
+      NANOARROW_RETURN_NOT_OK(ArrowArrayInitFromType(child_ptr, NANOARROW_TYPE_NA));
+      child_ptr->length     = child->size();
+      child_ptr->null_count = child->size();
+    } else {
+      NANOARROW_RETURN_NOT_OK(cudf::type_dispatcher(
+        child->type(), dispatch_to_arrow_device{}, *child, stream, mr, child_ptr));
+    }
+
+    child_ptr++;
+  }
+
+  NANOARROW_RETURN_NOT_OK(
+    ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_MINIMAL, nullptr));
+  ArrowArrayMove(tmp.get(), out);
+  return NANOARROW_OK;
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::list_view>(cudf::column& column,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr,
+                                                          ArrowArray* out)
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(ArrowArrayInitFromType(tmp.get(), NANOARROW_TYPE_LIST));
+  tmp->length     = column.size();
+  tmp->null_count = column.null_count();
+
+  auto contents = column.release();
+  if (contents.null_mask) {
+    NANOARROW_RETURN_NOT_OK(
+      set_buffer(std::move(contents.null_mask), kValidityBufferIdx, tmp.get()));
+  }
+  auto offsets_contents =
+    contents.children[cudf::lists_column_view::offsets_column_index]->release();
+  NANOARROW_RETURN_NOT_OK(set_buffer(std::move(offsets_contents.data), 1, tmp.get()));
+
+  std::unique_ptr<cudf::column>& child =
```

Comment:
I originally put the full type there instead of `auto` since it's not obvious that is the type on the right-hand side. Not a problem though, I'll switch it back to `auto`